### PR TITLE
Add support for reporting with custom severity

### PR DESCRIPTION
**Pull request by aaronccasanova**
### Short description

This PR introduces the ability to report warnings with custom rule severities:

```js
stylelint.utils.report({ ruleName, severity: 'warning' })
```

### Context

My team is working on a rule that is used for both linting and gathering metrics on warnings and have encountered a need to set custom rule severities per report. In short our rule dynamically categorizes warnings and reports with a modified rule name that we target on an external metrics dashboard e.g.

```js
stylelint.utils.report({ ruleName: `${ruleName}/colors`, ... })
stylelint.utils.report({ ruleName: `${ruleName}/layout`, ... })
// etc.
```

However, because we report dynamically created rule names the warning's `severity` is set to `undefined`.

### In summary

This small restructure and override mechanism gives us more programmatic control when handling warnings and from what I can tell doesn't affect the CLI or VS Code extension.

That said, I'm very interested to hear what folks think!

### CODE REVIEW

**Comment by ybiquitous** [comment]
@aaronccasanova Thanks for requesting the new feature and writing the code.

We usually implement a new feature after a discussion on an issue and reject direct pull requests without a discussion.

However, this feature request seems to make sense, and the code in the pull request is reliable because of enough test cases. Also, I believe adding the optional `severity` parameter should not break our ecosystem (3rd-party plugins).

But I'm not sure if I'm missing something about the new feature, so I've marked this *needs discussion* just in case.

**Comment by ybiquitous** [comment]
If there are no objections for a while, we can proceed.

**Comment by aaronccasanova** [comment]
Hello @ybiquitous 👋 Thanks for the quick response! I'd be happy to move this over to an issue and unpack more of the details. We have a pretty unique use case so I'd understand if the proposed changes don't align with the scope of the project and capabilities exposed to 3rd-party plugins.

As an aside, should we update the [Code contributions](https://stylelint.io/contributing/#code-contributions) or [Open a pull request](https://stylelint.io/contributing/#open-a-pull-request) section of the documentation to state the projects preference for creating issues before pull requests?

**Comment by ybiquitous** [comment]
Thanks for the proposal. But these cases aren't very common, so I don't think we need to update the contribution guide for now.

**Comment by aaronccasanova** [comment]
I encountered another use case that could leverage the proposed update. I'm interested in locally extending the behavior of the [declaration-property-unit-disallowed-list](https://stylelint.io/user-guide/rules/declaration-property-unit-disallowed-list/) rule to allow consumers to configure `severity` per property e.g.
```js
// Consumer config
module.exports = {
  plugins: ['@shopify/stylelint-polaris'],
  rules: {
    'stylelint-polaris/declaration-property-unit-disallowed-list': {
      'font-size': [['em', 'px'], {severity: 'error'}],
      '/^animation/': ['s', {severity: 'warning'}],
    },
  },
}
```

**Review by ybiquitous — APPROVED** [review]
I've removed the blocking labels since there are no objections.

Thanks for your patience. LGTM 👍🏼

**Comment by aaronccasanova** [comment]
Thanks @ybiquitous! Is there anything else I need to include in this PR (e.g. documentation) or is it ready to ship?

**Comment by ybiquitous** [comment]
No need to do anything else. I will merge this in a few days if there are no problems. And we will ship a new version, including this PR, within a few weeks at the latest.

**Comment by josser** [comment]
Not, sure but this change introduce some incompatibility with previous configuration. 

I.e, if I'm just update stylint, w/o changing my config: 

`{
  "extends": [
    "stylelint-config-standard",
    "stylelint-config-standard-scss",
    "stylelint-config-prettier",
    "stylelint-config-prettier-scss",
    "stylelint-prettier/recommended"
  ],
  "rules": {
    "selector-pseudo-element-no-unknown": [
      true,
      {
        "ignorePseudoElements": [
          "",
          "ng-deep"
        ]
      }
    ],
    "selector-class-pattern": "^.[a-z]([a-z0-9-]+)?(__([a-z0-9]+-?)+)?(--([a-z0-9]+-?)+){0,2}$",
    "function-no-unknown": [
      true,
      {
        "ignoreFunctions": [
          "-webkit-gradient",
          "from",
          "to"
        ]
      }
    ],
    "selector-type-no-unknown": null
  }
}
`

Im getting an error: 
Error: Unknown severity: "undefined"

**Comment by ybiquitous** [comment]
@josser Thanks for the report. Could you provide a minimal reproduction repository, including `.stylelintrc.json`, `package.json`, and CSS files?

**Comment by josser** [comment]
@ybiquitous sure, https://github.com/josser/stylelint-minimal-reproduce

**Comment by ybiquitous** [comment]
@josser Thank you so much! I could reproduce the bug, so I'll submit a pull request to fix it. Please be patient. Thanks again. 😄

**Comment by josser** [comment]
@ybiquitous Thank you! I'm not really good at css/scss, just updating packages :D

**Comment by ybiquitous** [comment]
This error also occurs in versions before v14.15.0, so this pull request seems not a cause:

```console
$ npx stylelint --version
14.14.1

$ npm run stylelint:check

> stylelint-error@1.0.0 stylelint:check
> stylelint '**/*.scss'

Error: Unknown severity: "undefined"
```

**Comment by josser** [comment]
Hm, but not reproducible in 14.6.0

**Comment by aaronccasanova** [comment]
> This error also occurs in versions before v14.15.0, so this pull request seems not a cause

Agreed, this seems unrelated. I pulled the example repo and ran the linter programmatically to gain more insight on the cause of the error e.g.
```js
// node lint-styles.mjs

import stylelint from 'stylelint'

const { results } = await stylelint.lint({
  files: '**/*.scss',
  configFile: '.stylelintrc',
})

console.log(results[0])

/*
Where results[0].warnings === [
  {
    line: 2,
    column: 3,
    endLine: 6,
    endColumn: 4,
    rule: undefined,
    severity: undefined,
    text: 'Cannot parse selector'
  }
]
*/
```

Searching through `stylelint` I see that error message referenced in two places ([parseSelector.js](https://github.com/stylelint/stylelint/blob/d1705d0563cf4e232a80d86bdbbeb433d474ccf6/lib/utils/parseSelector.js#L16) and [transformSelector.js](https://github.com/stylelint/stylelint/blob/d1705d0563cf4e232a80d86bdbbeb433d474ccf6/lib/utils/transformSelector.js#L15)). Do we just need to add a `severity: 'error'` property to the warning? Or are `stylelintType: 'parseError'` errors handled elsewhere?

**Comment by ybiquitous** [comment]
@aaronccasanova Thanks for digging into the problem. I found this error is caused by `stylelint-scss`, not `stylelint`. See below:

https://github.com/stylelint-scss/stylelint-scss/blob/1c7bc949ea9470a3f471b9670ba41c1946e7ce63/src/utils/parseSelector.js#L7

The `parseSelector` of `stylelint-scss` is different from our one: it misses the `stylelintType` property. So the error by `stylelint-scss` is not reported as a `parserError`:

https://github.com/stylelint/stylelint/blob/d1705d0563cf4e232a80d86bdbbeb433d474ccf6/lib/formatters/stringFormatter.js#L261

I found other bugs in the formatters, so I've opened PR #6480.

**Comment by ybiquitous** [comment]
PR #6480 is merged. The next version will fix the error. Thanks for the report and investigation! @josser @aaronccasanova
